### PR TITLE
Back-End Design Documentation

### PR DESCRIPTION
**Pull request by seanytak**
Address #4 for the Back-End Design Documentation

### CODE REVIEW

File: design_docs/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Shaka Streets Design Documentation
+
+## Front-End
+**Description:** The Front-End should be allow the user to select a set of locations that they want to visit. After the user input is complete, a single list of locations (of type Array[ShakaNode]) should be sent to the Back-End.
```
Comment (bmwfire) [inline design_docs/README.md]: "... should allow the user to ..."

Comment (bmwfire) [reply design_docs/README.md]: In addition, the front-end isn't allowing the user to select a set of locations. Rather, it is allowing the user to input a n amount of locations they want to visit.

File: design_docs/ShakaDistanceMatrix.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+## Interface
+
+### protocol ShakaDistanceMatrix
+
+**Methods:**
+- public addLocations(locations: Array[ShakaNode])
+  - **Parameters:**
+    - locations: Array[ShakaNode] -> New location(s) to be added
+  - **Description:**
+    - A list of locations will be added to the ShakaDistanceMatrix with their appropriate edge weights added for all entries as well
```
Comment (bmwfire) [inline design_docs/ShakaDistanceMatrix.md:13]: I may be wrong, but edge weights are only done when compared between two ShakaNodes, not from a single ShakaNode right? So a corresponding edge weight should be mapped to two ShakaNodes.

File: design_docs/ShakaDistanceMatrix.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    - Get a single distance (edge weight) from source to destination.
+- public removeLocations(locations: Array[ShakaNode])
+  - **Parameters:**
+    - locations: Array[ShakaNode] -> Location(s) to be removed
+  - **Description:**
+    - Remove locations from the ShakaDistanceMatrix
+- public updateEdgeWeights(locations: Array[ShakaNode])
+  - **Parameters:**
+    - locations: Array[ShakaNode] -> Location(s) to have their related edge weight(s) updated.
+
```
Comment (bmwfire) [inline design_docs/ShakaDistanceMatrix.md:43]: We should have a simple method that checks if the list of locations is empty meaning all inputted locations have been removed.

**Review by bmwfire — CHANGES_REQUESTED** [review]
Overall, this seems to be a good design doc for us. I've made a few changes that should be made before we merge this to our master repo. Nice job, Sean!

File: design_docs/ShakaDistanceMatrix.md (hunk 3 above)
Comment (seanytak) [reply design_docs/ShakaDistanceMatrix.md:43]: I agree. Added as 
- public isEmpty() => Boolean
  - **Returns:**
    - True if there are no more locations to visit.
  - **Description:**
    - Returns whether or not there are locations that still need to be visited. Implementation will need to somehow keep track of this.

File: design_docs/ShakaDistanceMatrix.md (hunk 2 above)
Comment (seanytak) [reply design_docs/ShakaDistanceMatrix.md:13]: Correct. There will be $V'$ locations added to a matrix of size $V^2$. The locations will turn the matrix into size $ (V' + V)^2 $ and should recompute all the distances upon addition to the matrix. I think the implementation design will guard against potential pitfalls will multiple calls to this method (defined using a 2d dictionary).

Comment (bmwfire) [reply design_docs/ShakaDistanceMatrix.md:13]: Ok cool

File: design_docs/ShakaDistanceMatrix.md (hunk 3 above)
Comment (bmwfire) [reply design_docs/ShakaDistanceMatrix.md:43]: Looks good!

**Review by bmwfire — APPROVED** [review]
This is a good design doc! Well done making the necessary changes. I approve of this PR. Good job Sean!

**Pull request merged**